File: CS50/Week 6/credit/credit.py
import re


def main() :
    number = get_string("Card number: ")
    card_type = get_type(number)
    if is_valid(card_type, number) == True : print(card_type)
    else : print("INVALID")


def get_string(string) :
    return input(string)


def get_type(number) :
    if re.fullmatch('^5[1-5][0-9]{14}$', number) != None : return 'MASTERCARD'
    elif re.fullmatch('^3[4|7][0-9]{13}$', number) != None : return 'AMEX'
    elif re.fullmatch('^4[0-9]{12}$', number) != None : return 'VISA'
    elif re.fullmatch('^4[0-9]{15}$', number) != None : return 'VISA'
    else : return 'INVALID'


def is_valid(card_type, card_number) :
    match card_type :
        case 'INVALID' : return False
        case 'VISA' : return True if digit_sum(card_number) % 10 == 0 else False
        case 'MASTERCARD' : return True if digit_sum(card_number) % 10 == 0 else False
        case 'AMEX' : return True if digit_sum(card_number) % 10 == 0 else False


def digit_sum(card_number) :
    card_number = int(card_number)
    sum_1 = 0
    sum_2 = 0
    while card_number > 0 :
        num_1 = card_number % 10
        sum_1 += num_1

        card_number //= 10

        num_2 = (card_number % 10) * 2
        num_2 = num_2 if num_2 <= 9 else num_2 % 10 + num_2 // 10
        sum_2 += num_2

        card_number //= 10
    return sum_1 + sum_2


main()

File: CS50/Week 1/Problem Set 1/credit/credit.c
#include <cs50.h>
#include <stdbool.h>
#include <stdio.h>

bool isLegit(long card_number);
int getNumberOfDigits(long card_number);
long getPowerOf10(int nr_of_digits);

int main(void)
{
	long card_number = get_long("Number: ");

	if(isLegit(card_number) == false)
	{
		printf("INVALID\n");
		return 1;
	}

	int nr_digits = getNumberOfDigits(card_number);

	if (nr_digits == 15 && (card_number / getPowerOf10(13) == 34 || (card_number / getPowerOf10(13) == 37)))
	{
		printf("AMEX\n");
		return 0;
	}
	
	if (nr_digits == 16)
	{
		if (card_number / getPowerOf10(14) >= 51 && card_number / getPowerOf10(14) <= 55)
		{
			printf("MASTERCARD\n");
			return 0;
		}
		else if (card_number / getPowerOf10(15) == 4)
		{
			printf("VISA\n");
			return 0;
		}
	}

	if (nr_digits == 13 && (card_number / getPowerOf10(12) == 4))
	{
		printf("VISA\n");
		return 0;
	}

	printf("INVALID\n");
	return 1;
}

bool isLegit(long card_number)
{
	long n = card_number;
	int sum_1 = 0;
	int sum_2 = 0;
	int sum_total = 0;
	while (n)
	{
		int digit = n % 100 / 10;
		
		digit *= 2;	

		if (digit > 9)
		{
			digit = digit % 10 + 1;
		}
		sum_1 += digit;
		sum_2 += n % 10;
		n /= 100;
	}

	sum_total = sum_1 + sum_2;

	if (sum_total % 10 == 0)
	{
		return true;
	}
	return false;
}

int getNumberOfDigits(long card_number)
{
	int n = 0;

	while(card_number)
	{
		n++;
		card_number /=10;
	}

	return n;
}

long getPowerOf10(int nr_of_digits)
{
	long output = 1;

	for (int i = 0; i < nr_of_digits; i++)
	{
		output *= 10;
	}

	return output;
}
